### ISSUE TITLE
Data loss when transferring sparse files between SSD and GPFS

### ISSUE THREAD

**Issue by tonyhutter** [IBM/CAST#918]
**Describe the bug**
If you have a sparse file, or a partially sparse file (with the sparse sections after the extents), then the file will not be transferred correctly using the BBAPI between the SSD and GPFS.  If you transfer a file from the SSD to GPFS that is completely sparse:
```
$ truncate -s 1M file1
```
You'll get a 0 byte file in GPFS.  If you transfer a file that's partially sparse:
```
$ echo "hello world" > file2 && truncate -s 1M file2
```
You'll get a file that's 65536 bytes in GPFS.  

Full details:
https://github.com/LLNL/scr/issues/163#issuecomment-631734186

**To Reproduce**
Replace the `randfile` command in  https://github.com/IBM/CAST/blob/master/bb/tests/src/test_basic_xfer.c#L233 with one of the commands above.

**Expected behavior**
I expect the destination files to be the same size as the source files (1MB).

**Environment (please complete the following information):**
 - Machine [lassen, LLNL]
 - Version [ibm-csm-1.6.0-2603, ibm-burstbuffer-1.6.2.3135]

**Additional context**
https://github.com/LLNL/scr/issues/163#issuecomment-631734186

**Issue Source:**
https://github.com/LLNL/scr/issues/163#issuecomment-631734186

**Comment by tgooding** [comment]
Looks like the apparent file size is incorrect but i'd surmise that all sparse data was transferred to GPFS intact.  BB looks at the allocated blocks and only transfers those.  However, if the file's last allocated block != last file block, then the file size (as reported by `ls`) won't match.  I think thats what you are seeing.  I'd agree that we should try to match the file metadata.  

@dlherms-ibm : can you look?

**Comment by dlherms-ibm** [comment]
Will do...

**Comment by tonyhutter** [comment]
Some additional stat and hexdump info:

```bash
$ echo "hello world" > hello
$ truncate -s 1M hello

$ ls -l hello 
-rw------- 1 hutter2 hutter2 1048576 May 21 09:08 hello

$ stat hello
  File: 'hello'
  Size: 1048576   	Blocks: 128        IO Block: 65536  regular file
Device: fd02h/64770d	Inode: 1027        Links: 1
Access: (0600/-rw-------)  Uid: (57592/ hutter2)   Gid: (57592/ hutter2)
Access: 2020-05-21 09:08:26.970249503 -0700
Modify: 2020-05-21 09:08:31.410053045 -0700
Change: 2020-05-21 09:08:31.410053045 -0700

# correctly hexdumps 1MB
$ hexdump hello
0000000 6568 6c6c 206f 6f77 6c72 0a64 0000 0000
0000010 0000 0000 0000 0000 0000 0000 0000 0000
*
0100000

[transfer file from SSD to GPFS]


$ ls -l /p/gpfs1/hutter2/hello 
-rw------- 1 hutter2 hutter2 65536 May 21 09:09 /p/gpfs1/hutter2/hello

$ stat /p/gpfs1/hutter2/hello 
  File: '/p/gpfs1/hutter2/hello'
  Size: 65536     	Blocks: 256        IO Block: 16777216 regular file
Device: 51h/81d	Inode: 340875      Links: 1
Access: (0600/-rw-------)  Uid: (57592/ hutter2)   Gid: (57592/ hutter2)
Access: 2020-05-21 09:09:50.650099000 -0700
Modify: 2020-05-21 09:09:50.744947347 -0700
Change: 2020-05-21 09:09:50.744947347 -0700
 Birth: -

# It only hexdumps 64k
$ hexdump /p/gpfs1/hutter2/hello 
0000000 6568 6c6c 206f 6f77 6c72 0a64 0000 0000
0000010 0000 0000 0000 0000 0000 0000 0000 0000
*
0010000
```

**Comment by tgooding** [comment]
(The SSD's XFS file system block size is set at 64 kiB (matching the POWER9's page size)  So your echo would have 1 allocated block starting at file byte 0)

**Comment by dlherms-ibm** [comment]
@tonyhutter Thanks for the detailed recreation steps.  I've been able to recreate this scenario.  We'll be working on a fix to get the metadata fixed up for a transferred sparse file having a trailing hole.

**Comment by tgooding** [comment]
This fix will be included in the forthcoming 1.8.0 release.

**Comment by tgooding** [comment]
1.8.0 has been out since June, closing/stale.  If this is still a problem, please reopen.  Thanks!
